---
title: "fig1e_louise"
format: html
editor: visual
---

## Replicating figure 1e from article

Heres some info on the figure (from article):\
*"RNA-Seq analysis of cultured cells and resulting density plots of gene expression levels. The density estimates of gene expression levels are shown for each cell type and for a combination over all cell types as indicated. The combined cell type was derived by extracting the largest expression value over all cell types. In all cell types, gene expression levels followed a binomial distribution (black). Gene expression filtered for RPKM \> 1 values are shown in blue"*

![](images/clipboard-2755963945.png)

Heres my script for the analysis

```{r}
library(tidyverse)
```

Notes:

-   Data in figure 1e is from supp table 3

-   Data is log2 transformed

-   There are 5 classes of column names:

    -   Oligodendrocytes (div1.1-1.3 + dic 2.5 1-3 + div 4 1-3)

    -   Adult microglia 1-3

    -   Atrocytes 1-3

    -   Cortical neurons div05 (1-3) + dig 10 (1-3)

    -   GeneName

```{r}
getwd()

#Going back 1 step to open the file
rna_data <- read_tsv("../data/dirty_data/supp_3.tsv", show_col_types = FALSE)

colnames(rna_data)
head(rna_data, 10)
```

Data is read. Steps now:

1\) Locate X and Y values (X=log2 RPKM, Y=density of these measurements, generated during ggplot)

2\) Choose specific columns to use

3\) Consider tidying the data more (from wide to long format (--\> make col names: Gene Name, Cell type, RPKM))

4\) Log2 transform values (calculate + add additional column for this)

5\) Make plot using ggplot2
